modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/validation/NameClashModelValidator.java

```diff
@@ -199,12 +220,34 @@ private void validateNameClashes(Collection<? extends Named>... collections) {
             i++;
           }
 
-          errorMessage.append(format(" which name is '%s'. Names should be unique", entry.getKey()));
+          errorMessage.append(format(" which alias name is '%s'. Names should be unique", entry.getKey()));
           throw new IllegalModelDefinitionException(errorMessage.toString());
         }
       });
     }
 
+    private void validateSingularizeNameClashesWithTopLevels() {
+      Map<String, Collection<TopLevelParameter>> singularClashes = topLevelParameters.keySet().stream()
+          .filter(k -> singularizedObjects.keySet().contains(k) && !topLevelParameters.get(k).isEmpty())
+          .collect(toMap(k -> k, topLevelParameters::get));
+
+      if (!singularClashes.isEmpty()) {
+        List<String> errorMessages = new ArrayList<>();
+        singularClashes.entrySet().forEach(e -> {
+          DescribedParameter reference = singularizedObjects.get(e.getKey());
+          e.getValue().forEach(tp -> errorMessages.add(format(
+                                                              "%s '%s' contains parameter '%s' that clash when singularized with parameter '%s' from '%s'",
+                                                              reference.parent.getDescription(), reference.parent.getName(),
+                                                              reference.getName(),
+                                                              tp.getName(), tp.ownerType)));
+        });
+
+        throw new IllegalModelDefinitionException(format("Extension '%s' contains %d parameters that clash when singularized. %s",
```

Comment:
"parameters that clash when singularized" is not a descriptive message. It should just say that it contains parameter names which clash when transformed into XML DSL language